### PR TITLE
Add ray tracing methods to LivingEntity

### PR DESCRIPTION
**Pull request by BillyGalbreath**
This method will return the Block a player is looking at while taking into consideration the AABB of each block in the path.

For example, you can look through the 1/4 space of air in a Stair block and get the block behind it instead of the Stair block you are looking past.

### CODE REVIEW

File: Spigot-Server-Patches/0359-Add-Player-getTrueTargetBlock.patch [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     }
+ 
++    // Paper start
++    public org.bukkit.block.Block getTrueTargetBlock(int reach) {
++        if (reach < 1 || reach > 120) {
++            throw new IllegalArgumentException("reach must be between 1-120");
++        }
++
++        BlockPosition start = new BlockPosition(locX, locY + getHeadHeight(), locZ);
++        org.bukkit.util.Vector dir = getBukkitEntity().getEyeLocation().getDirection().normalize().multiply(reach);
```
Comment (hugmanrique) [inline Spigot-Server-Patches/0359-Add-Player-getTrueTargetBlock.patch]: The normalize call isn't needed, getDirection already returns a unit vector (https://hub.spigotmc.org/javadocs/spigot/org/bukkit/Location.html#getDirection--)

**Comment by BillyGalbreath** [comment]
~~Dont pull yet. having some odd behavior now that I moved it to LivingEntity.~~

**Comment by BillyGalbreath** [comment]
Ok, this is good to go now.

Here is a video demonstrating this test case: https://youtu.be/HZ6ooXRpvrA

(blue particle is showing maxDistance (reach). red particles are outlining the **accurate** targeted Block. yellow particles are outlining the **non-accurate** (old API) targeted Block)

And here is my test plugin:

```
public class TestRayTrace extends JavaPlugin {
    private static final Set<Material> TRANSPARENT = new HashSet<>(Arrays.asList(Material.AIR, Material.CAVE_AIR, Material.VOID_AIR, Material.LAVA, Material.WATER));

    private int maxDistance = 5;

    @Override
    public void onEnable() {
        new BukkitRunnable() {
            public void run() {
                World world = Bukkit.getWorlds().get(0);
                if (world == null) {
                    return; // no worlds loaded
                }

                if (world.getPlayers().isEmpty()) {
                    return; // no players in this world
                }

                Player player = world.getPlayers().get(0);
                if (player == null) {
                    return; // no players online yet
                }

                // New API! Yay! \o/ ^_^
                Block targetedBlock = player.getTargetBlock(maxDistance);
                BlockFace targetedBlockFace = player.getTargetBlockFace(maxDistance);
                TargetBlockInfo targetBlockInfo = player.getTargetBlockInfo(maxDistance);

                // Old API! Booo! Hiiisss! >:(
                Block oldTargetedBlock = player.getTargetBlock(TRANSPARENT, maxDistance);
                System.out.println(oldTargetedBlock.getType());
                if (TRANSPARENT.contains(oldTargetedBlock.getType())) {
                    // remove ignored block to match
                    // new API for this demo
                    oldTargetedBlock = null;
                }

                spawnParticlesAtLimit(player, maxDistance, Color.BLUE);

                String targeted = "null";
                if (targetedBlock != null) {
                    spawnParticlesOnBlock(targetedBlock, Color.RED);
                    spawnParticlesOnHitBlockFace(player, maxDistance, Color.GREEN);
                    targeted = targetedBlock.getType().name();
                }

                String oldTarget = "null";
                if (oldTargetedBlock != null) {
                    spawnParticlesOnBlock(oldTargetedBlock, Color.YELLOW);
                    oldTarget = oldTargetedBlock.getType().name();
                }

                player.sendMessage(ChatColor.RED + targeted + " " + ChatColor.YELLOW + oldTarget);

                Scoreboard board = Bukkit.getScoreboardManager().getNewScoreboard();
                Objective objective = board.registerNewObjective("test", "dummy");
                objective.setDisplaySlot(DisplaySlot.SIDEBAR);
                objective.setDisplayName(ChatColor.RED + "Targets");
                Score score5 = objective.getScore(ChatColor.AQUA + "Block " + ChatColor.RED + (targetedBlock == null ? "null" : targetedBlock.getType().name()));
                score5.setScore(5);
                Score score4 = objective.getScore(ChatColor.AQUA + "BlockFace: " + ChatColor.GREEN + (targetedBlockFace == null ? "null" : targetedBlockFace.name()));
                score4.setScore(4);
                Score score3 = objective.getScore(ChatColor.AQUA + "Info Block " + ChatColor.RED + (targetBlockInfo == null ? "null" : targetBlockInfo.getBlock().getType().name()));
                score3.setScore(3);
                Score score2 = objective.getScore(ChatColor.AQUA + "Info BlockFace: " + ChatColor.GREEN + (targetBlockInfo == null ? "null" : targetBlockInfo.getBlockFace().name()));
                score2.setScore(2);
                Score score1 = objective.getScore(ChatColor.AQUA + "Old Block " + ChatColor.YELLOW + (oldTargetedBlock == null ? "null" : oldTargetedBlock.getType().name()));
                score1.setScore(1);
                player.setScoreboard(board);
            }
        }.runTaskTimer(this, 1, 1);
    }

    @Override
    public boolean onCommand(CommandSender sender, Command cmd, String label, String[] args) {
        int num;
        try {
            num = Integer.parseInt(args[0]);
        } catch (NumberFormatException e) {
            sender.sendMessage("Not a number");
            return true;
        }

        if (num < 1 || num > 120) {
            sender.sendMessage("Outside range (1-120)");
            return true;
        }

        maxDistance = num;
        sender.sendMessage("maxDistance: " + maxDistance);
        return true;
    }

    private void spawnParticlesOnBlock(Block block, Color color) {
        block.getWorld().spawnParticle(Particle.REDSTONE, block.getX(), block.getY(), block.getZ(), 1, 0, 0, 0, new Particle.DustOptions(color, 1));
        block.getWorld().spawnParticle(Particle.REDSTONE, block.getX() + 1, block.getY(), block.getZ(), 1, 0, 0, 0, new Particle.DustOptions(color, 1));
        block.getWorld().spawnParticle(Particle.REDSTONE, block.getX() + 1, block.getY(), block.getZ() + 1, 1, 0, 0, 0, new Particle.DustOptions(color, 1));
        block.getWorld().spawnParticle(Particle.REDSTONE, block.getX(), block.getY(), block.getZ() + 1, 1, 0, 0, 0, new Particle.DustOptions(color, 1));
        block.getWorld().spawnParticle(Particle.REDSTONE, block.getX(), block.getY() + 1, block.getZ(), 1, 0, 0, 0, new Particle.DustOptions(color, 1));
        block.getWorld().spawnParticle(Particle.REDSTONE, block.getX() + 1, block.getY() + 1, block.getZ(), 1, 0, 0, 0, new Particle.DustOptions(color, 1));
        block.getWorld().spawnParticle(Particle.REDSTONE, block.getX() + 1, block.getY() + 1, block.getZ() + 1, 1, 0, 0, 0, new Particle.DustOptions(color, 1));
        block.getWorld().spawnParticle(Particle.REDSTONE, block.getX(), block.getY() + 1, block.getZ() + 1, 1, 0, 0, 0, new Particle.DustOptions(color, 1));
    }

    private void spawnParticlesAtLimit(Player player, int maxDistance, Color color) {
        net.minecraft.server.v1_13_R2.Entity entity = ((CraftPlayer) player).getHandle();

        Vec3D start = new Vec3D(entity.locX, entity.locY + entity.getHeadHeight(), entity.locZ);
        org.bukkit.util.Vector dir = player.getLocation().getDirection().multiply(maxDistance);
        Vec3D end = new Vec3D(start.x + dir.getX(), start.y + dir.getY(), start.z + dir.getZ());

        player.getWorld().spawnParticle(Particle.REDSTONE, end.x, end.y, end.z, 1, 0, 0, 0, new Particle.DustOptions(color, 1));
    }

    private void spawnParticlesOnHitBlockFace(Player player, int maxDistance, Color color) {
        net.minecraft.server.v1_13_R2.Entity entity = ((CraftPlayer) player).getHandle();

        Vec3D start = new Vec3D(entity.locX, entity.locY + entity.getHeadHeight(), entity.locZ);
        org.bukkit.util.Vector dir = player.getLocation().getDirection().multiply(maxDistance);
        Vec3D end = new Vec3D(start.x + dir.getX(), start.y + dir.getY(), start.z + dir.getZ());

        WorldServer world = ((CraftWorld) player.getWorld()).getHandle();

        MovingObjectPosition rayTrace = world.rayTrace(start, end);

        if (rayTrace == null) {
            return;
        }

        Block block = CraftBlock.at(world, rayTrace.a().shift(rayTrace.direction));

        spawnParticlesOnBlock(block, color);
    }
}
```

**Comment by chickeneer** [comment]
It would be interesting to know if the `direction` variable in MovingObjectPosition, returns the BlockFace of the block you are looking at. If so, that has the potential to be useful, as where a block would be placed when staring at a given block.
(no complaints about the PR as is)

**Comment by BillyGalbreath** [comment]
@chickeneer It does. ;)

This screenshot is outlining the neighboring block

image

```
Block shiftedBlock = CraftBlock.at(world, rayTrace.a().shift(rayTrace.direction));
spawnParticlesOnBlock(shiftedBlock, Color.GREEN);
```

Notice, my cursor is on the side of the sign ;)

**Comment by aikar** [comment]
Maybe getBlockFacingTargetBlock() too?

or TargetBlockInfo getTargetBlock(int);  that has a wrapper for the target + direction?

though maybe make that a separate method, TargetBlockInfo getTargetBlockWithFacing(int)

File: Spigot-Server-Patches/0359-Add-ray-tracing-methods-to-LivingEntity.patch [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
++import org.bukkit.block.BlockFace;
+ import org.bukkit.craftbukkit.CraftWorld;
+ import org.bukkit.craftbukkit.util.Waitable;
+ import org.spigotmc.AsyncCatcher;
+@@ -330,4 +332,36 @@ public final class MCUtil {
+     public static World getNMSWorld(@Nonnull org.bukkit.entity.Entity entity) {
+         return getNMSWorld(entity.getWorld());
+     }
++
++    public static FluidCollisionOption getNMSFluidCollisionOption(TargetBlockInfo.FluidMode fluidMode) {
```
Comment (aikar) [inline Spigot-Server-Patches/0359-Add-ray-tracing-methods-to-LivingEntity.patch]: Can you move this to the mcutil patch

File: Spigot-Server-Patches/0359-Add-ray-tracing-methods-to-LivingEntity.patch [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
++        if (fluidMode == TargetBlockInfo.FluidMode.SOURCE_ONLY) {
++            return FluidCollisionOption.SOURCE_ONLY;
++        }
++        if (fluidMode == TargetBlockInfo.FluidMode.ALWAYS) {
++            return FluidCollisionOption.ALWAYS;
++        }
++        return null;
++    }
++
++    public static BlockFace toBukkitBlockFace(EnumDirection enumDirection) {
```
Comment (aikar) [inline Spigot-Server-Patches/0359-Add-ray-tracing-methods-to-LivingEntity.patch]: and this

File: Spigot-Server-Patches/0359-Add-ray-tracing-methods-to-LivingEntity.patch [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
++import org.bukkit.craftbukkit.block.CraftBlock;
+ import org.bukkit.craftbukkit.inventory.CraftEntityEquipment;
+ import org.bukkit.craftbukkit.inventory.CraftItemStack;
+ import org.bukkit.craftbukkit.potion.CraftPotionUtil;
+@@ -169,6 +175,35 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
+         return blocks.get(0);
+     }
+ 
++    // Paper start
++    public Block getTargetBlock(int maxDistance) {
```
Comment (aikar) [inline Spigot-Server-Patches/0359-Add-ray-tracing-methods-to-LivingEntity.patch]: these defaults can be done at the -API level

**Review by aikar — CHANGES_REQUESTED** [review]

**Review by aikar — APPROVED** [review]

**Pull request merged**